Komorebi: Build system replacement.

Kangie opened the pull request:
Replaced CMake with Meson for Komorebi.

It's easier to read and maintain, and is the preferred option based on previous discussions. This replaces my previous PR for Komorebi FHS compliance - Meson is handling that for us now, and it *should* be respecting prefixes properly at this point. 

As always, please test before you accept / merge; It's 0430 here and while I've tested I can't be sure that I haven't introduced any new bugs.

Also note that there are a few hardcoded paths in the Komorebi code, so if you change your prefix there may be some unexpected behaviour.
see: [Here](https://github.com/Komorebi-Fork/komorebi/blob/master/src/OnScreen/WallpapersSelector.vala#L27) and [here](https://github.com/Komorebi-Fork/komorebi/blob/master/src/OnScreen/Thumbnail.vala#L28). If someone has the time to look into that I'd appreciate it - I'm kinda beat.

Ev1lbl0w commented:
I can't review this properly for the next week or so, but after a quick "skim" it looks good.

I saw that you bumped some versions, should we then prepare for a release after this PR?

Ev1lbl0w commented:
CC @FFY00 from previous discussion

Kangie commented:
@Ev1lbl0w, Yeah, I've bumped a few versions in preparation for a release; I don't think that I've covered everything - once this is merged I plan to submit another PR to finalise all the version-bump-y bits and pieces, and do file headers and the like. The ones that I updated were targets of opportunity, or places that it made sense to me. e.g. meson build system change should be at least a minor version bump. 

I'm still not sure what to do about the hardcoded file paths, though. Any thoughts?

Ev1lbl0w commented:
> I'm still not sure what to do about the hardcoded file paths, though. Any thoughts?

I think we should change them, even if they don't do anything, just for the sake of consistency.

On that topic, the code should also be using meson variables like `datadir` instead of hardcoded paths, because the prefix might be different.

For that, it's probably better to have a single file, or use `Utilities.vala` so it holds the base paths, and change the code to use that instead. This way, meson only has to modify 1 file, and we also prevent any dangling paths in the code.

FFY00 commented on data/Other/postinst:
```diff
@@ -190,10 +190,10 @@ Height=0
 
-done	
\ No newline at end of file
+done
```
This whole script is something that should be rethought and probably removed, but that can come in a future step.

Kangie replied:
Agreed.

FFY00 commented on data/meson.build:
```diff
@@ -0,0 +1,36 @@
+
+install_subdir('Wallpapers', install_dir: 'share' / meson.project_name(), strip_directory: true)
+install_subdir('Fonts', install_dir: 'share/fonts', strip_directory: true)
+install_subdir('Icons', install_dir: 'share/pixmaps' / meson.project_name(), strip_directory: true)
```
Use `'share' / 'pixmaps' / meson.project_name()`, etc.

FFY00 commented on meson.build:
```diff
@@ -0,0 +1,38 @@
+project('komorebi', 'vala', 'c', version: '2.2.0', meson_version: '>=0.49', default_options: ['prefix=/usr'])
```
Please put the kw arguments in new lines.

The default prefix should absolutely be `/usr/local`, which is what meson defaults to. You should never manually install anything to the `/usr` prefix, that is where your distro puts files :smile:

FFY00 commented on src/OnScreen/AssetActor.vala:
```diff
@@ -54,7 +54,7 @@ namespace Komorebi.OnScreen {
                 assetHeight = screenHeight;
 
-            var assetPath = @"/System/Resources/Komorebi/$wallpaperName/assets.png";
+            var assetPath = @"/usr/share/komorebi/$wallpaperName/assets.png";
```
Please pass `-DPREFIX=$prefix` in meson instead of hardcoding the path here.

FFY00 commented on src/meson.build:
```diff
@@ -0,0 +1,18 @@
+sources = files('Main.vala',
```
nitpick: `Main.vala` can go to the line below

FFY00 commented on wallpaper-creator/Main.vala:
```diff
@@ -59,7 +59,7 @@ namespace WallpaperCreator {
 
         if(args[1] == "--version" || args[1] == "version") {
-            print("Version: 1.1 - Summit\nCreated by: Abraham Masri @cheesecakefuo\n\n");
+            print("Version: 1.2\nMaintained by: Komorebi Team\n\n");
```
Similar to prefix, meson should pass the version.

Arham4 commented:
If I recall correctly, the squashing can be done automatically when merging into the repo. I know GitLab has this feature, can't recall if GitHub does, however.

Arham4 commented:
Confirmed: 
![image](https://user-images.githubusercontent.com/2886217/87608007-647c9f80-c6c4-11ea-9095-58be9869e45f.png)s

Kangie commented:
@Arham4 , I think that rather than squashing the build system changes into one commit isn't what @FFY00 had in mind - There's a few different changes in this PR (though they're mostly related). 

My read is that we'd prefer a few commits: 

1. Build System -> Meson (also includes moving extra/wallpaper-creator up by one level)
2. CMake Removed
3. Style Changes (fixing whitespace / adding newlines at end of file).
4. Documentation updates. 

So I'm squashing my current commits down to 4 more logical commits, and then (post review) rebasing those commits and merging, to make it easier to bisect for bugs, rather than one massive 'I've done a bunch of stuff, removed CMake, and added Meson'.

@FFY00 did I understand your git history comment correctly?

After a few days on it I think that I've worked out how you're *supposed* to pass configuration from your build system to Vala, so I'll get to work on replacing hardcoded paths, versions (etc), and once that's doing what it should I'll fix the commit history.

Arham4 commented:
True. Point 3 and 4 could maybe be omitted, though.

On Wed, Jul 15, 2020, 21:35 Matt Jolly <notifications@github.com> wrote:

> @Arham4 <https://github.com/Arham4> , I think that rather than squashing
> the build system changes into one commit isn't what @FFY00
> <https://github.com/FFY00> had in mind - There's a few different changes
> in this PR (though they're mostly related).
>
> My read is that we'd prefer a few commits:
>
>    1. Build System -> Meson (also includes moving extra/wallpaper-creator
>    up by one level)
>    2. CMake Removed
>    3. Style Changes (fixing whitespace / adding newlines at end of file).
>    4. Documentation updates.
>
> So I'm squashing my current commits down to 4 more logical commits, and
> then (post review) rebasing those commits and merging, to make it easier to
> bisect for bugs, rather than one massive 'I've done a bunch of stuff,
> removed CMake, and added Meson'.
>
> @FFY00 <https://github.com/FFY00> did I understand your git history
> comment correctly?
>
> After a few days on it I think that I've worked out how you're *supposed*
> to pass configuration from your build system to Vala, so I'll get to work
> on replacing hardcoded paths, versions (etc), and once that's doing what it
> should I'll fix the commit history.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Komorebi-Fork/komorebi/pull/14#issuecomment-659121221>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAWAUSIZIPQDWSSUBC7EENLR3ZRQJANCNFSM4OXNOCJQ>
> .
>

Kangie commented:
OK, this one just kept snowballing.

All of the images used within Komorebi are now packaged into a gresource that's packaged with the binary - the only images that we're now installing (aside from wallpapers...) are the Icons, the rest are accessed via the resource '/org/komorebiteam/komorebi'. There's probably room for some slight optimisation - all of the images are svg (so xml), could probably strip out whitespace and compress them, but I honestly don't see it making a big difference, it's on the order of 25Kbit.

Meson should now configure a new 'config.vala' file with version, app name (for ease of renaming), and the path to the wallpaper directory (datadir). It's a little hacky, but I'd rather get the build system doing what it's supposed to be doing and implement that in a better way down the track.

Something to note: The Main app and Wallpaper Creator now have synced versions - I couldn't see a good reason to do it differently, and the version info is populated from Meson anyway so it can all be easily updated in one place.

I've done a bit of testing and (as far as I can tell) everything seems to behave as I'd expect it to, things are installed to whatever prefix you specify, and it generally seems to work.

@Komorebi-Fork/komorebi-fork-core any feedback before we merge?

Ev1lbl0w commented on src/OnScreen/WallpapersSelector.vala:
```diff
@@ -55,20 +55,22 @@ namespace Komorebi.OnScreen {
 				thumbnailsList.remove(thumbnail);
 
-			File wallpapersFolder = File.new_for_path("/System/Resources/Komorebi");
+			File wallpapersFolder = File.new_for_path(@"$package_datadir");
```
Since this is just a single variable, you can just supply `package_datadir` directly.

Kangie replied:
As it turns out, you can do exactly that. This is a great example of why I try not to code into the wee hours...

I've got a changed staged locally to fix this, because why not, but I'll wait to see if there's other feedback to bundle into the commit before I push.

Kangie commented:
@FFY00, Will you have a chance to review this anytime soon?

FFY00 commented on meson.build:
```diff
@@ -0,0 +1,55 @@
+pkgdatadir = join_paths(prefix, datadir, meson.project_name())
+
+# Configure Komorebi
+conf = configuration_data()
+conf.set_quoted('PACKAGE_NAME', meson.project_name())
+conf.set_quoted('PACKAGE_VERSION', meson.project_version())
+conf.set_quoted('PKGDATADIR', pkgdatadir)
+komorebiconfiguration = configure_file(input: 'config.vala.in', output: 'config.vala', configuration: conf)
```
I would have done this by injecting the defines via the compiler but fair enough :grin:

Kangie replied:
The Vala guys pointed me at this method for what we're doing, but if you can point me at some resources for doing it that way that I'd love to implement it more elegantly!

FFY00 replied:
No need. This is perfectly fine.

FFY00 commented on data/meson.build:
```diff
@@ -0,0 +1,40 @@
+  )
+endforeach
+
+install_subdir('Wallpapers', install_dir: 'share' / meson.project_name(), strip_directory: true)
+install_subdir('Fonts', install_dir: 'share' / 'fonts', strip_directory: true)
```
`get_option('datadir')` instead of harcoding `share`

FFY00 commented on data/meson.build:
```diff
@@ -0,0 +1,40 @@
+
+# Install only the app icons
+install_data('resources/komorebi.svg', 'resources/wallpaper_creator.svg',
+  install_dir: 'share' / 'pixmaps' / meson.project_name()
```
Same

FFY00 commented on data/meson.build:
```diff
@@ -0,0 +1,40 @@
+# Setup configuration data for .desktop entries
+desktop_conf = configuration_data()
+desktop_conf.set('bindir', join_paths(prefix, bindir))
+desktop_conf.set('datadir', join_paths(prefix,datadir))
```
nitpick: space after the comma

I would also use `/`, but you can keep this if you want.

Kangie commented:
> Sorry for the delay!

No worries - You have other things on your plate. Thanks for the feedback!

I thought I replaced all those join_paths() at some point in the process. Obviously missed a few.
